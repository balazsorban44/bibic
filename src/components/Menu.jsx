import React, {
  lazy, Suspense, Component
} from 'react'
import {Link as RouteLink, withRouter, useHistory} from 'react-router-dom'
import {Link} from 'react-scroll'
import bibic from '../assets/icons/bibic.png'
import logo from '../assets/icons/logo.png'
import Fade from "react-reveal/Fade"
import {BASE_URL} from '../utils/constants'
import {withTranslation, useTranslation} from 'react-i18next'
import LanguageSelector from './LanguageSelector'

const RoomMenu = lazy(() => import("./RoomMenu"))


export class Menu extends Component {

  state = {
    isMenuOpen: false,
    isScrolled: false,
    isRoomMenuOpen: true,
    menu: [
      {name: "intro", to:"bemutatkozas"},
      {name: "facilities", to:"szolgaltatasok"},
      {name: "rooms", to:"szobak"},
      {name: "prices", to:"arak"},
      {
        name: "menu", to:"etelek", component: RouteLink
      },
      {
        name: "events", to:"rendezvenyek", component: RouteLink
      },
      {name: "feedback", to:"visszajelzesek"},
      {name: "contact", to:"kapcsolat"}
    ]
  }

  componentDidMount() {
    window.addEventListener("scroll", this.isScrolled, false)
    window.addEventListener("resize", this.isResized, false)
    this.isScrolled()
    this.isResized()
    if (window.innerWidth > 768) this.setState({isRoomMenuOpen: false})
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.isScrolled, false)
    window.removeEventListener("resize", this.isResized, false)
  }

  isScrolled = () => this.setState({isScrolled: window.pageYOffset >= 100})

  isResized = () => this.setState({width: window.innerWidth})

  handleShowRoomMenu = () => this.setState({isRoomMenuOpen: true})

  handleHideRoomMenu = () => this.setState({isRoomMenuOpen: false})

  handleMenuToggle = () => this.setState(({isMenuOpen}) => ({isMenuOpen: !isMenuOpen}))

  handleHideMenu = () => this.setState({isMenuOpen: false, isRoomMenuOpen: false})

  render() {
    const {
      isMenuOpen, isScrolled, isRoomMenuOpen, width, menu
    } = this.state
    const isBigScreen = width > 768
    return (
      <Fade down>
        <div
          className={`menu ${isScrolled ? "menu-scrolled" : ""} ${isMenuOpen ? "" : "menu-hidden"}`}
        >
          <Fade right>
            <div
              className={`hamburger ${isMenuOpen ? "hamburger-clicked" : ""}`}
              onClick={this.handleMenuToggle}
            ><span/><span/><span/></div>
          </Fade>
          <a
            className={`menu-logo ${isScrolled ? "menu-logo-scrolled" : ""}`}
            href={BASE_URL}
          >
            <picture>
              <source
                media="(min-aspect-ratio: 7/6) and (min-width: 1024px)"
                srcSet={logo}
              />
              <img
                alt="logo"
                src={bibic}
              />
            </picture>
          </a>
          <nav>
            <Fade
              cascade
              down
            >
              <ul>
                {menu.map(({
                  name, to, component: Component=Link, offset
                }) =>
                  <li key={to}>
                    <Component
                      href={`#${to}`}
                      offset={isBigScreen ? offset !== undefined ? offset : -64 : 0}
                      onClick={this.handleHideMenu}
                      onMouseEnter={() => to==="szobak" && this.handleShowRoomMenu()}
                      smooth={Component === Link ? true : undefined}
                      to={to}
                    >
                      {this.props.t(`menu.${name}`)}
                    </Component>
                  </li>
                )}
                <LanguageSelector/>
              </ul>
            </Fade>
          </nav>
          <Suspense fallback={null}>
            <ul className={`room-menu ${isRoomMenuOpen ? "room-menu-show" : ""}`}>
              <RoomMenu
                isBigScreen={isBigScreen}
                onClick={this.handleHideMenu}
              />
            </ul>
          </Suspense>
        </div>
      </Fade>
    )
  }
}

export default withTranslation()(Menu)


export const BackMenu = () => {
  const {goBack, location: {search}} = useHistory()
  const customReturn = new URLSearchParams(search).get("backTo")
  const [t] = useTranslation()
  const title = customReturn || t("menu.homepage")

  if (customReturn) {
    return (
      <button
        className="back-to-home"
        onClick={goBack}
        title={title}
      >
        <Title title={title}/>
      </button>
    )
  } else {
    return (
      <RouteLink
        className="back-to-home"
        title={title}
        to="/"
      >
        <Title title={title}/>
      </RouteLink>
    )
  }
}

const Title = ({title}) => {
  const [t] = useTranslation()
  return (
    <>
      {"‚Üê"}
      <span>
        {t("menu.back-to", {title})}
      </span>
    </>
  )
}