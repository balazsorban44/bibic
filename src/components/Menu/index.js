import React, {Component} from 'react'
import {Link} from 'react-scroll'


export default class Menu extends Component {
  render() {
    return (
      <header>
        <nav className="hidden">
          <ul className="hidden" id="menu">
            <li><Link offset={-64} smooth to="bemutatkozas">Bemutatkozás</Link></li>
            <li><Link offset={-64} smooth to="szolgaltatasok">Szolgáltatások</Link></li>
            <li><Link offset={-64} smooth to="szobak">Szobák</Link></li>
            <li><Link offset={-64} smooth to="specialitasok">Specialitások</Link></li>
            <li><Link offset={-64} smooth to="arak">Árak</Link></li>
            <li><Link offset={-64} smooth to="galeria">Galéria</Link></li>
            <li><Link offset={-64} smooth to="kapcsolat">Kapcsolat</Link></li>
          </ul>
          <ul id="szoba-lista">
            <Link offset={-64} smooth to="szoba-1">1</Link>
            <Link offset={-64} smooth to="szoba-2">2</Link>
            <Link offset={-64} smooth to="szoba-3">3</Link>
            <Link offset={-64} smooth to="szoba-4">4</Link>
            <Link offset={-64} smooth to="szoba-5">5</Link>
            <Link offset={-64} smooth to="szoba-6">6</Link>
          </ul>
        </nav>
      </header>
    )
  }
}