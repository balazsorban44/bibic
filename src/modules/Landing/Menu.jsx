import React, {lazy, Suspense, useState} from "react"
import {Link as RouteLink} from "react-router-dom"
import bibic from "assets/icons/bibic.min.png"
import logo from "assets/icons/logo.min.png"
import Fade from "react-reveal/Fade"
import {useTranslation} from "react-i18next"
import withLazy from "components/withLazy"
import {useScrolled, useSize} from "hooks"

const LanguageSelector =
  () => withLazy(lazy(() => import("components/LanguageSelector")))(undefined, {fallback: null})

const RoomMenu = lazy(() => import("./RoomMenu"))
const Link = lazy(() => import("react-scroll/modules/components/Link"))


const menu = [
  {to:"bemutatkozas", name: "intro"},
  {to:"szolgaltatasok", name: "facilities"},
  {to:"szobak", name: "rooms"},
  {to:"arak", name: "prices"},
  {to:"etelek", name: "foods", component: RouteLink},
  {to:"rendezvenyek", name: "events", component: RouteLink},
  {to:"visszajelzesek", name: "feedbacks"},
  {to:"kapcsolat", name: "contact"}
]


function useMenu() {
  const [menu, setMenu] = useState(false)
  const [roomMenu, setRoomMenu] = useState(false)

  return {
    menuOpen: menu,
    roomMenuOpen: roomMenu,
    hideMenu: () => {
      setMenu(false)
      setRoomMenu(false)
    },
    showRoomMenu: () => setRoomMenu(true),
    toggleMenu: () => setMenu(menu => !menu)
  }
}


export default function Menu() {
  const [t] = useTranslation("common")

  const scrolled = useScrolled({offset: 100})

  const {
    menuOpen, roomMenuOpen,
    hideMenu, showRoomMenu, toggleMenu
  } = useMenu()

  const {width} = useSize()
  const mdUp = width > 768


  return (
    <Fade down>
      <div className={`menu ${scrolled ? "menu-scrolled " : ""}${menuOpen ? "" : "menu-hidden"}`}>
        <Fade right>
          <div
            className={`hamburger ${menuOpen ? "hamburger-clicked" : ""}`}
            onClick={toggleMenu}
          ><span/><span/><span/></div>
        </Fade>
        <div
          className="menu-logo"
          href={process.env.REACT_APP_BASE_URL}
        >
          <picture>
            <source
              media="(min-aspect-ratio: 7/6) and (min-width: 1024px)"
              srcSet={logo}
            />
            <img
              alt="logo"
              src={bibic}
            />
          </picture>
        </div>
        <nav>
          <Suspense fallback={null}>
            <Fade
              cascade
              down
            >
              <ul className="nav-menu">
                {menu.map(({name, to, component: Component=Link, offset=-64}) =>
                  <li key={to}>
                    <Component
                      offset={mdUp ? offset : 0}
                      onClick={hideMenu}
                      onMouseEnter={() => to==="szobak" && showRoomMenu()}
                      smooth={Component === Link || undefined}
                      to={to}
                    >
                      {t(`menu.${name}`)}
                    </Component>
                  </li>
                )}
                <LanguageSelector/>
              </ul>
            </Fade>
          </Suspense>
        </nav>
        <Suspense fallback={null}>
          <RoomMenu
            mdUp={mdUp}
            onClick={hideMenu}
            open={roomMenuOpen}
          />
        </Suspense>
      </div>
    </Fade>
  )
}