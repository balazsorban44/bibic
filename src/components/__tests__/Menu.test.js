import {Menu, BackMenu} from "../Menu"
import {MemoryRouter} from "react-router"

describe("Menu component", () => {
  const props = {rooms: [{id: 1}]}
  const wrapper = shallow(
    <MemoryRouter>
      <Menu {...props}/>
    </MemoryRouter>
  )

  it("renders correctly", () => {
    expect(wrapper).toHaveLength(1)
  })

  describe("Navigation menu", () => {
    const navMenu = wrapper.dive().dive().find("ul").first()

    it.skip("has 8 elements", () => {
      expect(navMenu.find("li").length).toBe(8)
    })

    it.skip("clicking element hides menu", () => {
      wrapper.setState({isMenuOpen: true, isRoomMenuOpen: true})
      navMenu.find("li").first().simulate("click")
      expect(wrapper.state("isMenuOpen")).toBe(false)
      expect(wrapper.state("isRoomMenuOpen")).toBe(false)
    })
  })

  it.skip("handleShowRoomMenu", () => {
    wrapper.childAt(0).childAt(0).instance().handleShowRoomMenu()
    expect(wrapper.childAt(0).childAt(0).state("isRoomMenuOpen")).toBe(true)
  })

  it.skip("handleHideRoomMenu", () => {
    wrapper.childAt(0).childAt(0).instance().handleHideRoomMenu()
    expect(wrapper.childAt(0).childAt(0).state("isRoomMenuOpen")).toBe(false)
  })

  it.skip("handleHideMenu", () => {
    wrapper.childAt(0).childAt(0).instance().handleHideMenu()
    expect(wrapper.childAt(0).childAt(0).state("isRoomMenuOpen")).toBe(false)
    expect(wrapper.childAt(0).childAt(0).state("isMenuOpen")).toBe(false)
  })

  it.skip("handleMenuToggle", () => {
    wrapper.childAt(0).childAt(0).setState({isMenuOpen: false})
    wrapper.childAt(0).childAt(0).instance().handleMenuToggle()
    expect(wrapper.childAt(0).childAt(0).state("isMenuOpen")).toBe(true)
    wrapper.childAt(0).childAt(0).instance().handleMenuToggle()
    expect(wrapper.childAt(0).childAt(0).state("isMenuOpen")).toBe(false)
  })
})


describe("BackMenu component", () => {
  const wrapper = shallow(<BackMenu/>)

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot()
  })
})