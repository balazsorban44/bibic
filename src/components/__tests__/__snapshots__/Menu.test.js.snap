// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackMenu component renders correctly 1`] = `ShallowWrapper {}`;
